### PR TITLE
Add bounded-collections crate

### PR DESCRIPTION
**Pull request by KiChjang**
This new crate contains the bounded types that we see in Substrate, and traits that are used to support the implementation of such types.

Related to paritytech/polkadot#6572.

### CODE REVIEW

**Review by kianenigma — APPROVED** [review]

**Comment by KiChjang** [comment]
@mrcnski I originally thought of putting the `Const*` types into `primitive-types`, but I read the crate description and it said "Primitive types shared between Substrate and Ethereum", and since the `Const*` types and `Get` traits aren't in Ethereum, I ultimately decided against it. Do you still think it's a better fit to move these types and traits there though? What other places do you have in mind instead?

**Comment by mrcnski** [comment]
@KiChjang Gotcha, I just think it would be odd to depend upon `bounded-collections` if I wanted to use e.g. `ConstBool`. I don't know the dependency stack well enough yet to offer an existing alternative. But maybe there could be another new crate here in `parity-common` called `traits`?

**Comment by KiChjang** [comment]
So, the name of the crate needs to be unique, as they will be published on crates.io, thus I can't just call it `traits` and be done with it. I've searched in Substrate and found out that `primitive-types` is depended upon by `sp-core`, `sp-arithmetic` and `sp-runtime-interface`.

What would likely happen next is that `bounded-collections` will be re-exported from `sp-core`, and for crates that are critical in memory usage, we'd just depend directly upon `bounded-collections` rather than going through `sp-core`.

**Comment by KiChjang** [comment]
me gusta core-types, let's do so.

**Review by mrcnski — APPROVED** [review]
Awesome! I like the new name, but think someone else should approve it too.

**Comment by niklasad1** [comment]
I think `core-types` is too generic as we already got `primitive_types` the distinction between is blurry IMO (by crate names).
I prefer `bounded-collections` as you did initially 

I requested a review from @ordian he knows this crate the best :)

**Review by gilescope — APPROVED** [review]

File: core-types/src/weak_bounded_vec.rs [hunk 1]
```diff
@@ -0,0 +1,518 @@
...
+		// plus each item in the slice:
+		// See: https://docs.substrate.io/reference/scale-codec/
+		codec::Compact(S::get())
+			.encoded_size()
+			.saturating_add(Self::bound().saturating_mul(T::max_encoded_len()))
+	}
+}
+
+#[cfg(test)]
+pub mod test {
```
Comment (bkchr) [inline core-types/src/weak_bounded_vec.rs:448]:
```suggestion
mod test {
```

File: core-types/src/bounded_btree_map.rs [hunk 2]
```diff
@@ -0,0 +1,618 @@
...
+		if self.len() > Bound::get() as usize {
+			Err("iterator length too big")
+		} else {
+			Ok(BoundedBTreeMap::<K, V, Bound>::unchecked_from(self.collect::<BTreeMap<K, V>>()))
+		}
+	}
+}
+
+#[cfg(test)]
+pub mod test {
```
Comment (bkchr) [inline core-types/src/bounded_btree_map.rs:393]:
```suggestion
mod test {
```

File: core-types/src/bounded_btree_set.rs [hunk 3]
```diff
@@ -0,0 +1,479 @@
...
+		if self.len() > Bound::get() as usize {
+			Err("iterator length too big")
+		} else {
+			Ok(BoundedBTreeSet::<T, Bound>::unchecked_from(self.collect::<BTreeSet<T>>()))
+		}
+	}
+}
+
+#[cfg(test)]
+pub mod test {
```
Comment (bkchr) [inline core-types/src/bounded_btree_set.rs:323]:
```suggestion
mod test {
```

File: core-types/src/bounded_vec.rs [hunk 4]
```diff
@@ -0,0 +1,1268 @@
...
+		if self.len() > Bound::get() as usize {
+			Err("iterator length too big")
+		} else {
+			Ok(BoundedVec::<T, Bound>::unchecked_from(self.collect::<Vec<T>>()))
+		}
+	}
+}
+
+#[cfg(all(test, feature = "std"))]
+pub mod test {
```
Comment (bkchr) [inline core-types/src/bounded_vec.rs:903]:
```suggestion
mod test {
```

File: core-types/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,22 @@
+[package]
+name = "core-types"
```
Comment (bkchr) [inline core-types/Cargo.toml:2]: Please go with `bounded-collections` or whatever was the name before. This is just tooo generic.

**Comment by ordian** [comment]
I do not quite get why this crate should live in parity-common and not in Substrate? It looks quite stand-alone and non of the workspace crates here depend on it (or is it going to change?).

The historical purpose of parity-common was to share code between substrate and ethereum client code bases and nowadays it's substrate and frontier/evm.

**Review by ordian — APPROVED** [review]
That being said, I've probably missed something and don't want be a blocker for merging this PR.

**Comment by KiChjang** [comment]
> It looks quite stand-alone and non of the workspace crates here depend on it (or is it going to change?)

Yes, we're essentially pulling out some common types that are going to be used in Substrate and later on the XCM crate when it lives in its own repository, independent of both Polkadot and Substrate.

**Pull request merged**